contracts/WooCrossChainRouter.sol

```diff
@@ -1,372 +1,446 @@
-        wooppQuoteTokens[110] = address(0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8); // Arbitrum: usdc
-        wooppQuoteTokens[111] = address(0x7F5c764cBc14f9669B88837ca1490cCa17c31607); // Optimism: usdc
-        wooppQuoteTokens[112] = address(0x04068DA6C83AFCFA0e13ba15A6696662335D5B75); // ftm_wooPP: usdc
-    }
+        sglChainId = _sglChainId;
 
-    function setWooppQuoteTokens(uint16 _chainId, address _token) public onlyOwner {
-        wooppQuoteTokens[_chainId] = _token;
+        _initSGETHs();
+        _initSGPoolIds();
     }
 
-    /*
-    https://stargateprotocol.gitbook.io/stargate/developers/contract-addresses/mainnet
-        - Chain ID : Chain -
-        1: Ether
-        2: BSC (BNB Chain)
-        6: Avalanche
-        9: Polygon
-        10: Arbitrum
-        11: Optimism
-        12: Fantom
-    */
-    function setWooCrossChainRouter(uint16 _chainId, address _wooCrossRouter) external onlyOwner {
-        require(_wooCrossRouter != address(0), "WooCrossChainRouter: !wooCrossRouter");
-        wooCrossRouters[_chainId] = _wooCrossRouter;
-    }
+    /* ----- Functions ----- */
 
-    function setStargateRouter(address _stargateRouter) external onlyOwner {
-        require(_stargateRouter != address(0), "WooCrossChainRouter: !stargateRouter");
-        stargateRouter = IStargateRouter(_stargateRouter);
-    }
-
-    function setWooRouter(address _wooRouter) external onlyOwner {
-        wooRouter = IWooRouterV2(_wooRouter);
-    }
-
-    function setBridgeSlippage(uint256 _bridgeSlippage) external onlyOwner {
-        require(_bridgeSlippage <= 10000, "WooCrossChainRouter: !_bridgeSlippage");
-        bridgeSlippage = _bridgeSlippage;
-    }
-
-    function setDstGasForSwapCall(uint256 _dstGasForSwapCall) external onlyOwner {
-        dstGasForSwapCall = _dstGasForSwapCall;
-    }
+    function crossSwap(
+        uint256 refId,
+        address payable to,
+        SrcInfos memory srcInfos,
+        DstInfos memory dstInfos
+    ) external payable override {
+        require(srcInfos.fromToken != address(0), "WooCrossChainRouterV2: !srcInfos.fromToken");
+        require(dstInfos.toToken != address(0), "WooCrossChainRouterV2: !dstInfos.toToken");
+        require(to != address(0), "WooCrossChainRouterV2: !to");
+
+        uint256 srcPoolId = sgPoolIds[sglChainId][srcInfos.bridgeToken];
+        require(srcPoolId > 0, "WooCrossChainRouterV2: !srcInfos.bridgeToken");
+
+        uint256 dstPoolId = sgPoolIds[dstInfos.chainId][dstInfos.bridgeToken];
+        require(dstPoolId > 0, "WooCrossChainRouterV2: !dstInfos.bridgeToken");
+
+        address msgSender = _msgSender();
+        uint256 msgValue = msg.value;
+        uint256 bridgeAmount;
 
-    function setDstGasForNoSwapCall(uint256 _dstGasForNoSwapCall) external onlyOwner {
-        dstGasForNoSwapCall = _dstGasForNoSwapCall;
-    }
+        {
+            if (srcInfos.bridgeToken == ETH_PLACEHOLDER_ADDR) {
```

Comment:
Is this allowed to be set as "ETH_PLACEHOLDER_ADDR" ?

Reply:
`bridgeToken` could be `USDC/USDT/BUSD`, not only `ETH_PLACEHOLDER_ADDR`. Have a look on this function `_initSGPoolIds`.